service-compiler/src/main/java/io/techery/analytics/compiler/ActionHelperGenerator.java

```diff
@@ -79,8 +80,15 @@ private CodeBlock makeGetActionCodeBlock(AnalyticActionClass actionClass) {
       final CodeBlock.Builder builder = CodeBlock.builder();
 
       if (actionClass.containsActionPathParam) {
-         builder.addStatement("return $S.replace($S, action.$L)",
-               actionClass.action, ACTION_PATH_PARAM, actionClass.actionPartFieldName);
+         builder.addStatement("$T $L = $S", String.class, GENERATED_ACTION_KEY_VAR_NAME, actionClass.action);
+         for (KeyPathEntity keyPathEntity : actionClass.keyPathEntities) {
```

Comment:
in terms of performance, it's no so good to have multiple lines of string instances with replace method -> so better to replace actual string with it's parts and `+` signs

Reply:
Looking at effort needed to implement this I'd say that it does not worth it.
Please approve that this is really a blocker

Reply:
it's not a blocker, an improvement